src/test/java/com/ita/if103java/ims/service/WarehouseServiceImplTest.java

```diff
@@ -0,0 +1,454 @@
+        address = new Address("Ukraine", "Kyiv", "Stusa, 5", "77000", 48F, 50F);
+        addressDto = new AddressDto(1L, "Ukraine", "Kyiv", "Stusa, 5", "77000", 48F, 50F);
+    }
+
+    @Test
+    public void add_successFlow() {
+        int quantity = 2;
+        int level = 0;
+        int maxQuantity = userDetails.getAccountType().getMaxWarehouses();
+        Warehouse warehouseCreate = new Warehouse(1L, "WarehouseTop", "auto parts", 0, false, null, 1L, null, true);
+        warehouseDto.setParentID(null);
+        Long accountId = topWarehouseDto.getAccountID();
+        assertNull(topWarehouseDto.getParentID());
+        when(warehouseDao.findQuantityOfWarehousesByAccountId(accountId)).thenReturn(quantity);
+        assertTrue(quantity < maxQuantity);
+        when(warehouseDao.create(warehouseDtoMapper.toEntity(warehouseDto))).thenReturn(warehouseCreate);
+        when(addressDtoMapper.toEntity(warehouseDto.getAddressDto())).thenReturn(address);
+        assertTrue(warehouseCreate.isTopLevel());
+        when(addressDao.createWarehouseAddress(warehouseCreate.getId(), address)).thenReturn(address);
+        when(addressDtoMapper.toDto(address)).thenReturn(addressDto);
+
+        Event event = new Event("Warehouse created " +
+            warehouse.getName(), accountId,
+            warehouse.getId(), 1L, EventName.WAREHOUSE_CREATED, 2L);
+        doNothing().when(eventService).create(event);
+//      verify(eventService, times(1)).create(event);
+//         //form a return DTO
+        when(warehouseDtoMapper.toDto(warehouseCreate)).thenReturn(topWarehouseDto);
+        topWarehouseDto.setAddressDto(addressDto);
+
+    }
+
+    @Test
+    public void addWarehouse_testMaxWarehouseNumberReached() {
+        int quantity = 3;
+        warehouseDto.setParentID(null);
+        Long accountId = topWarehouseDto.getAccountID();
+        assertNull(topWarehouseDto.getParentID());
+        when(warehouseDao.findQuantityOfWarehousesByAccountId(1L)).thenReturn(quantity);
+
+        MaxWarehousesLimitReachedException exception = assertThrows(MaxWarehousesLimitReachedException.class, () -> {
+            warehouseService.add(warehouseDto, userDetails);
+        });
+        assertEquals("The maximum number of warehouses has been reached for this" + "{accountId = " + accountId + "}", exception.getMessage());
+    }
+
+    @Test
+    public void addWarehouse_testMaxDepthReached() {
+        Warehouse warehouse = new Warehouse(1L, "Bottom", "auto parts", 20, true, 5L, 1L, 4L, true);
+        int level = 4;
+        warehouseDto.setParentID(5L);
+        accountId = userDetails.getUser().getAccountId();
+        int maxDepth = userDetails.getAccountType().getMaxWarehouseDepth();
+        assertNotNull(warehouseDto.getParentID());
+        when(warehouseDao.findLevelByParentID(warehouseDto.getParentID())).thenReturn(level);
+        assertFalse(level < maxDepth);
```

Comment:
this test test nothing, there is no invocation of the warehouseService.add
Fix or remove